src/component/handlers/edit/editOnBeforeInputAndroid.js

```diff
@@ -0,0 +1,198 @@
+  );
+  return EditorState.push(editorState, contentState, 'insert-characters');
+}
+
+const logChanges = (editor, mutation) => {
+  const blocksBefore = editor._latestEditorState.getCurrentContent().getBlockMap();
+  const selectionBefore = editor._latestEditorState.getSelection();
+  mutation();
+  const blocksAfter = editor._latestEditorState.getCurrentContent().getBlockMap();
+  const selectionAfter = editor._latestEditorState.getSelection();
+
+  if (!blocksBefore.equals(blocksAfter)) {
+    console.log(
+      'Blocks Changed',
+      Object.values(blocksBefore.toJS()).map(v => `${v.key}: ${v.text}`),
+      Object.values(blocksAfter.toJS()).map(v => `${v.key}: ${v.text}`),
+    );
+  }
+  if (!selectionBefore.equals(selectionAfter)) {
+    console.log('Selection Changed', selectionBefore.toJS(), selectionAfter.toJS());
+  }
+};
+
+/**
+ * When `onBeforeInput` executes, the browser is attempting to insert a
+ * character into the editor. Apply this character data to the document
+ */
+function editOnBeforeInputAndroid(editor: DraftEditor, e: InputEvent): void {
+  if (e.isComposing && !e.cancelable) {
+    // Allow normal browser before for any composition events,
+    // TODO: This is the hard part, we bail out here so that we
+    // don't double
+    return;
+  }
+
+  const staticRanges = e.getTargetRanges();
+  const { inputType, data } = e;
+  const editorState = editor._latestEditorState;
+
+  const editorNode = ReactDOM.findDOMNode(editor.refs.editorContainer);
+
+  invariant(editorNode, 'Missing editorNode');
+  invariant(
+    editorNode.firstChild instanceof HTMLElement,
+    'editorNode.firstChild is not an HTMLElement',
+  );
+
+  const hasAffectedRanges =
+    staticRanges && Array.isArray(staticRanges) && staticRanges.length > 0;
+
+  // Compute
+  let affectedSelection = undefined;
+  if (hasAffectedRanges) {
+    const [affectedRange] = staticRanges;
+    affectedSelection = getDraftEditorSelectionWithNodes(
+      editorState,
+      editorNode.firstChild,
+      affectedRange.startContainer,
+      affectedRange.startOffset,
+      affectedRange.endContainer,
+      affectedRange.endOffset,
+    );
+  } else {
+    // If there is no affected range, we're just going to use the current native selection
+    // (implies we're about to do something additive)
+    const currentDraftSelectionFromCurrentNativeSelection = getDraftEditorSelection(
+      editorState,
+      editorNode.firstChild,
+    );
+    affectedSelection = currentDraftSelectionFromCurrentNativeSelection;
+  }
+
+  const editorStateWithCorrectSelection = EditorState.acceptSelection(
+    editorState,
+    affectedSelection.selectionState,
+  );
+  const chars = data;
+
+  switch (inputType) {
+    // Allowing insertCompositionText to pass through.
+    case 'insertCompositionText':
+      return;
+
+    case 'insertText':
+    case 'insertFromComposition':
+      if (!chars) {
+        return;
+      }
+      const newEditorState = replaceText(
```

Comment:
I think it's technically a no-no to declare variables in switch statements, but we can wrap it in an IIFE or add braces to the switch statement.